In [1]:
!pip install geopandas

  Using cached geopandas-0.10.2-py2.py3-none-any.whl (1.0 MB)
  Using cached Shapely-1.8.0-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.whl (1.1 MB)
  Using cached pyproj-3.2.1-cp39-cp39-manylinux2010_x86_64.whl (6.3 MB)
  Using cached Fiona-1.8.20-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.whl (15.4 MB)
  Using cached cligj-0.7.2-py3-none-any.whl (7.1 kB)
  Using cached click_plugins-1.1.1-py2.py3-none-any.whl (7.5 kB)
  Using cached munch-2.5.0-py2.py3-none-any.whl (10 kB)


In [2]:
import calitp
from calitp.tables import tbl
from calitp import get_engine
from siuba import *

import pandas as pd
import numpy as np
import geopandas as gpd

In [3]:
engine = get_engine()
connection = engine.connect()

In [4]:
from calitp import query_sql

In [5]:
query_sql("select * from views.mst_ridership")

participant_id              littlepay_transaction_id device_id  \
0                  sacrt  50e5c916-3f88-48f9-ae2d-3adbdc70a1c0  17F313D3   
1                  sacrt  b20c4c5a-5bf7-412d-8aa1-f767e9fdaf57  17F313D3   
2                  sacrt  f18e39e2-3eb5-4d48-8009-e25565b9fae9  17F313D3   
3                  sacrt  ee2c45ee-9d9b-43f9-ba33-aa8dba13f967  17F313D3   
4                  sacrt  047b4251-321f-44ee-9e3f-b6d602783bbd  17F313D3   
...                  ...                                   ...       ...   
31172  clean-air-express  dc20e795-261b-41c6-98a2-a620c85fa229     61005   
31173  clean-air-express  157d4863-6ce6-4147-88d0-ba1f3e434ff3     61005   
31174  clean-air-express  04d94e51-0384-46ff-9c3f-0722cb8191cd     61005   
31175  clean-air-express  26da56c2-185d-445a-b9fe-f7e24b449cca     61005   
31176  clean-air-express  60846f92-8e6c-49e8-9f29-23264d56384a     61005   

                            micropayment_id    type transaction_outcome  \
0      f8136ba0-5f66-410c-aa2e-736ae14fbbf6  single               allow   
1      698c2b29-14a3-4a25-b20a-f836b4cd9412  single               allow   
2      269d7920-9ae3-491c-be7c-f4061d0b9658  single               allow   
3      fe8a234f-26d4-4929-9c4e-9c2c5e89e9c6  single               allow   
4      54ef76b3-ead4-4cdc-90e4-5b5fe98cb624  single               allow   
...                                     ...     ...                 ...   
31172  fae9164c-9de8-414a-95d4-28234e12f7ae  single               allow   
31173  ebe1340a-b104-4cb1-abfb-ec44343da5cf  single               allow   
31174  5eaa4fad-53b5-4e1d-a012-a5d83832ecf2  single               allow   
31175  701f04b8-737d-4daf-b6c1-cd3255bc8165  single               allow   
31176  be1b0cc0-f74c-4acb-b3c7-7fe566219b63  single               allow   

      transaction_date_time_utc transaction_date_time_pacific location_id  \
0      2021-09-28T17:26:29.000Z           2021-09-28 10:26:29        7018   
1      2021-10-25T15:29:17.000Z           2021-10-25 08:29:17        7026   
2      2021-10-25T15:35:42.000Z           2021-10-25 08:35:42        7029   
3      2021-09-28T16:03:41.000Z           2021-09-28 09:03:41        7043   
4      2021-09-30T22:58:45.000Z           2021-09-30 15:58:45        7051   
...                         ...                           ...         ...   
31172  2021-08-05T00:20:09.000Z           2021-08-04 17:20:09     1679031   
31173  2021-08-11T12:55:24.000Z           2021-08-11 05:55:24     1689413   
31174  2021-08-10T12:55:57.000Z           2021-08-10 05:55:57     1689413   
31175  2021-08-09T12:57:28.000Z           2021-08-09 05:57:28     1689413   
31176  2021-08-09T12:53:45.000Z           2021-08-09 05:53:45     1689413   

                   location_name route_id            latitude  \
0          Broadway Station (EB)  Route Z   38.55976104736328   
1            Florin Station (WB)  Route Z   38.49806594848633   
2      City College Station (WB)  Route Z   38.54241943359375   
3        Royal Oaks Station (EB)  Route Z   38.60477828979492   
4       29th Street Station (EB)  Route Z    38.5640754699707   
...                          ...      ...                 ...   
31172              CtgePebCast O     None  34.428949499999995   
31173                     Lompoc     None   34.63738316666667   
31174                     Lompoc     None  34.637450333333334   
31175                     Lompoc     None  34.637398499999996   
31176                     Lompoc     None   34.63740433333333   

                 longitude          vehicle_id route_long_name  \
0       -121.4881820678711                 101            None   
1      -121.47150421142578                 101            None   
2      -121.48528289794922                 101            None   
3      -121.44329833984375                 101            None   
4      -121.46965789794922                 101            None   
...                    ...                 ...             ...   
31172  -119.72366933333332  

In [6]:
df = pd.read_sql_query("select * from views.mst_ridership", connection)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31177 entries, 0 to 31176
Data columns (total 19 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   participant_id                 31177 non-null  object        
 1   littlepay_transaction_id       31177 non-null  object        
 2   device_id                      31177 non-null  object        
 3   micropayment_id                31177 non-null  object        
 4   type                           31177 non-null  object        
 5   transaction_outcome            31177 non-null  object        
 6   transaction_date_time_utc      31177 non-null  object        
 7   transaction_date_time_pacific  31177 non-null  datetime64[ns]
 8   location_id                    31177 non-null  object        
 9   location_name                  31177 non-null  object        
 10  route_id                       27089 non-null  object        
 11  latitude       

In [8]:
df.head()

participant_id              littlepay_transaction_id device_id  \
0          sacrt  50e5c916-3f88-48f9-ae2d-3adbdc70a1c0  17F313D3   
1          sacrt  b20c4c5a-5bf7-412d-8aa1-f767e9fdaf57  17F313D3   
2          sacrt  f18e39e2-3eb5-4d48-8009-e25565b9fae9  17F313D3   
3          sacrt  ee2c45ee-9d9b-43f9-ba33-aa8dba13f967  17F313D3   
4          sacrt  047b4251-321f-44ee-9e3f-b6d602783bbd  17F313D3   

                        micropayment_id    type transaction_outcome  \
0  f8136ba0-5f66-410c-aa2e-736ae14fbbf6  single               allow   
1  698c2b29-14a3-4a25-b20a-f836b4cd9412  single               allow   
2  269d7920-9ae3-491c-be7c-f4061d0b9658  single               allow   
3  fe8a234f-26d4-4929-9c4e-9c2c5e89e9c6  single               allow   
4  54ef76b3-ead4-4cdc-90e4-5b5fe98cb624  single               allow   

  transaction_date_time_utc transaction_date_time_pacific location_id  \
0  2021-09-28T17:26:29.000Z           2021-09-28 10:26:29        7018   
1  2021-10-25T15:29:17.000Z           2021-10-25 08:29:17        7026   
2  2021-10-25T15:35:42.000Z           2021-10-25 08:35:42        7029   
3  2021-09-28T16:03:41.000Z           2021-09-28 09:03:41        7043   
4  2021-09-30T22:58:45.000Z           2021-09-30 15:58:45        7051   

               location_name route_id           latitude            longitude  \
0      Broadway Station (EB)  Route Z  38.55976104736328   -121.4881820678711   
1        Florin Station (WB)  Route Z  38.49806594848633  -121.47150421142578   
2  City College Station (WB)  Route Z  38.54241943359375  -121.48528289794922   
3    Royal Oaks Station (EB)  Route Z  38.60477828979492  -121.44329833984375   
4   29th Street Station (EB)  Route Z   38.5640754699707  -121.46965789794922   

  vehicle_id route_long_name route_short_name charge_type  charge_amount  \
0        101            None             None   flat_fare            1.0   
1        101            None             None   flat_fare            2.5   
2        101            None             None   flat_fare            0.0   
3        101            None             None   flat_fare            1.0   
4        101            None             None   flat_fare            1.0   

   nominal_amount  
0             1.0  
1             2.5  
2             2.5  
3             1.0  
4             1.0

Route Type	Regular Fare	Discount Fare
Local	    $1.50	        $.75
Primary	    $2.50	        $1.25
Regional	$3.50	        $1.75
Commuter	$12.00	        $6.00

Distance	                Regular Contactless Fare
Up to 2.7 miles	            $1.50
Between 2.7 and 14 miles	$2.50
14 – 50 miles	            $3.50
50 – 77 miles	            $7.00
77 miles or more	        $10.00

In [9]:
df['charge_type'].value_counts()

complete_variable_fare    22502
flat_fare                  8675
Name: charge_type, dtype: int64

21534 fares are variable, 8264 are flat

In [10]:
df['charge_amount'].max()

10.0

No one was charged the 12 regular commuter fare; capped out at 10.

In [11]:
df['charge_amount'].min()

0.0

In [12]:
df['route_id'].value_counts()

020        4726
902        3047
018        2308
002        2149
901        1997
041        1890
Route Z    1817
023        1674
024        1606
001         856
044         689
049         598
045         595
075         546
029         394
011         358
022         285
094         248
028         236
016         218
093         169
003         112
095          97
021          95
061          60
24X          51
091          48
026          43
025          36
019          26
048          18
084          17
12X          15
007          14
11           13
092          11
20           10
6             6
047           2
042           2
4             2
25            2
15X           2
16            1
Name: route_id, dtype: int64

** are the 384 0.00 fares actually no fares or missing data?

In [13]:
df['charge_amount'].value_counts()

1.50     11860
2.50      9124
3.50      5258
1.75      2250
6.00      1111
4.00       518
1.00       502
0.00       401
10.00       80
8.50        14
2.00        12
0.50        11
0.75        11
7.50        10
7.00         4
5.00         4
1.25         3
3.00         2
6.50         2
Name: charge_amount, dtype: int64

*df['charge_amount'].equals(df['nominal_amount'])

*comparison_column = np.where(df["charge_amount"] == df["nominal_amount"], True, False)
*print(comparison_column)

*df["equal"] = comparison_column
*print(df)

In [14]:
df['higher_fare'] = df['charge_amount'].gt(df['nominal_amount'])
df['lower_fare'] = df['charge_amount'].lt(df['nominal_amount'])
df['equal_fare'] = df['charge_amount'].eq(df['nominal_amount'])

In [15]:
df.head()

participant_id              littlepay_transaction_id device_id  \
0          sacrt  50e5c916-3f88-48f9-ae2d-3adbdc70a1c0  17F313D3   
1          sacrt  b20c4c5a-5bf7-412d-8aa1-f767e9fdaf57  17F313D3   
2          sacrt  f18e39e2-3eb5-4d48-8009-e25565b9fae9  17F313D3   
3          sacrt  ee2c45ee-9d9b-43f9-ba33-aa8dba13f967  17F313D3   
4          sacrt  047b4251-321f-44ee-9e3f-b6d602783bbd  17F313D3   

                        micropayment_id    type transaction_outcome  \
0  f8136ba0-5f66-410c-aa2e-736ae14fbbf6  single               allow   
1  698c2b29-14a3-4a25-b20a-f836b4cd9412  single               allow   
2  269d7920-9ae3-491c-be7c-f4061d0b9658  single               allow   
3  fe8a234f-26d4-4929-9c4e-9c2c5e89e9c6  single               allow   
4  54ef76b3-ead4-4cdc-90e4-5b5fe98cb624  single               allow   

  transaction_date_time_utc transaction_date_time_pacific location_id  \
0  2021-09-28T17:26:29.000Z           2021-09-28 10:26:29        7018   
1  2021-10-25T15:29:17.000Z           2021-10-25 08:29:17        7026   
2  2021-10-25T15:35:42.000Z           2021-10-25 08:35:42        7029   
3  2021-09-28T16:03:41.000Z           2021-09-28 09:03:41        7043   
4  2021-09-30T22:58:45.000Z           2021-09-30 15:58:45        7051   

               location_name  ...            longitude vehicle_id  \
0      Broadway Station (EB)  ...   -121.4881820678711        101   
1        Florin Station (WB)  ...  -121.47150421142578        101   
2  City College Station (WB)  ...  -121.48528289794922        101   
3    Royal Oaks Station (EB)  ...  -121.44329833984375        101   
4   29th Street Station (EB)  ...  -121.46965789794922        101   

  route_long_name route_short_name charge_type charge_amount nominal_amount  \
0            None             None   flat_fare           1.0            1.0   
1            None             None   flat_fare           2.5            2.5   
2            None             None   flat_fare           0.0            2.5   
3            None             None   flat_fare           1.0            1.0   
4            None             None   flat_fare           1.0            1.0   

   higher_fare  lower_fare  equal_fare  
0        False       False        True  
1        False       False        True  
2        False        True       False  
3        False       False        True  
4        False       False        True  

[5 rows x 22 columns]

In [16]:
df['equal_fare'].value_counts()

True     30313
False      864
Name: equal_fare, dtype: int64

In [17]:
df['higher_fare'].value_counts()

False    31177
Name: higher_fare, dtype: int64

In [18]:
df['lower_fare'].value_counts()

False    30313
True       864
Name: lower_fare, dtype: int64

So this means that fare capping worked, yes? No one was charged a higher amount than the nominal fare? 

In [19]:
df['charge_type'].value_counts()

complete_variable_fare    22502
flat_fare                  8675
Name: charge_type, dtype: int64

In [20]:
(
tbl.views.mst_ridership()
    >> filter(_.charge_amount < _.nominal_amount, _.charge_type == "flat_fare")
    >> summarize(n_fares = _.charge_type.count()
                )
)

# Source: lazy query
# DB Conn: Engine(bigquery://cal-itp-data-infra/?maximum_bytes_billed=5000000000)
# Preview:
   n_fares
0      712
# .. may have more rows

In [21]:
(
tbl.views.mst_ridership()
    >> filter(_.charge_amount < _.nominal_amount, _.charge_type == "complete_variable_fare")
    >> summarize(n_fares = _.charge_type.count()
                )
)

# Source: lazy query
# DB Conn: Engine(bigquery://cal-itp-data-infra/?maximum_bytes_billed=5000000000)
# Preview:
   n_fares
0      152
# .. may have more rows

Of the 820 people who were charged less than the nominal amount, 670 paid a flat fare (which means they are using a discount fare), and 150 used contactless. 

In [25]:
#test

In [27]:
#test

In [28]:
#test